packages/backend/src/modules/tracked-txs/modules/l2-costs/L2CostsAggregatorIndexer.ts

```diff
@@ -0,0 +1,230 @@
+import { assert, AssetId, clampRangeToDay, UnixTime } from '@l2beat/shared-pure'
+
+import {
+  ManagedChildIndexer,
+  type ManagedChildIndexerOptions,
+} from '../../../../tools/uif/ManagedChildIndexer'
+import type { PriceRepository } from '../../../tvl/repositories/PriceRepository'
+import type {
+  AggregatedL2CostsRecord,
+  AggregatedL2CostsRepository,
+} from './repositories/AggregatedL2CostsRepository'
+import type {
+  L2CostsRecord,
+  L2CostsRecordWithProjectId,
+  L2CostsRepository,
+} from './repositories/L2CostsRepository'
+
+// Amount of gas required for a basic tx
+const OVERHEAD = 21_000
+
+export interface L2CostsAggregatorIndexerDeps
+  extends Omit<ManagedChildIndexerOptions, 'name'> {
+  l2CostsRepository: L2CostsRepository
+  aggregatedL2CostsRepository: AggregatedL2CostsRepository
+  priceRepository: PriceRepository
+}
+
+export class L2CostsAggregatorIndexer extends ManagedChildIndexer {
+  constructor(private readonly $: L2CostsAggregatorIndexerDeps) {
+    super({ ...$, name: 'l2_costs_aggregator' })
+  }
+
+  override async update(from: number, to: number): Promise<number> {
+    const [shiftedFrom, shiftedTo] = this.shift(from, to)
+
+    if (shiftedFrom.equals(shiftedTo)) {
+      // there's nothing to sync
+      return to
+    }
+
+    const costs = await this.$.l2CostsRepository.getWithProjectIdByTimeRange([
+      shiftedFrom,
+      shiftedTo,
+    ])
+
+    const ethPrices = await this.$.priceRepository.findByTimestampRange(
+      AssetId.ETH,
+      shiftedFrom,
+      shiftedTo,
+    )
+
+    const aggregated = this.aggregate(costs, ethPrices)
+    await this.$.aggregatedL2CostsRepository.addMany(aggregated)
+    this.logger.info('Aggregated L2 costs', {
+      count: aggregated.length,
+    })
+
+    return shiftedTo.add(1, 'seconds').toNumber()
+  }
+
+  override async invalidate(targetHeight: number): Promise<number> {
+    //TODO: Check if targetHeight is inclusive
```

Comment:
To resolve

Reply:
Resolved: it should be exclusive (`targetHeight` refers to the last entry that should remain)

Reply:
Then we should change `deleteFrom (>=)` to `deleteAfter (>)`

Reply:
Done